pkg/rbac/sharedData.go

```diff
@@ -0,0 +1,143 @@
+	"github.com/stolostron/search-v2-api/pkg/config"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog/v2"
+)
+
+// Cache data shared across all users.
+type SharedData struct {
+	// These are the data fields.
+	csResources []resource // Cluster-scoped resources (ie. Node, ManagedCluster)
+	namespaces  []string
+
+	// These are internal objects to track the state of the cache.
+	csErr       error      // Capture errors retrieving cluster-scoped resources.
+	csLock      sync.Mutex // Locks the csResources map while updating it.
+	csUpdatedAt time.Time  // Time when cluster-scoped data was last updated.
+	nsErr       error      // Capture errors retrieving namespaces.
+	nsLock      sync.Mutex // Locks the namespaces array while updating it.
+	nsUpdatedAt time.Time  // Time when namespaces data was last updated.
+}
+
+type resource struct {
+	apigroup string
+	kind     string
+}
+
+func (cache *Cache) ClusterScopedResources(ctx context.Context) ([]resource, error) {
+	clusterScoped, err := cache.shared.getClusterScopedResources(cache, ctx)
+	return clusterScoped, err
+}
+
+func (shared *SharedData) getClusterScopedResources(cache *Cache, ctx context.Context) ([]resource,
+	error) {
+
+	//lock to prevent checking more than one at a time
+	shared.csLock.Lock()
+	defer shared.csLock.Unlock()
+	if shared.csResources != nil &&
+		time.Now().Before(shared.csUpdatedAt.Add(time.Duration(config.Cfg.SharedCacheTTL)*time.Millisecond)) {
+		klog.V(8).Info("Using cluster scoped resources from cache.")
+		return shared.csResources, shared.csErr
+	}
+	klog.V(6).Info("Querying database for cluster-scoped resources.")
+	shared.csErr = nil // Clear previous errors.
+
+	// if data not in cache or expired
+	schemaTable := goqu.S("search").Table("resources")
+	ds := goqu.From(schemaTable)
+
+	//"SELECT DISTINCT(data->>apigroup, data->>kind) FROM search.resources WHERE
+	// cluster='local-cluster' AND namespace=NULL"
+	query, _, err := ds.SelectDistinct(goqu.COALESCE(goqu.L(`"data"->>'apigroup'`), "").As("apigroup"),
+		goqu.COALESCE(goqu.L(`"data"->>'kind'`), "").As("kind")).
+		Where(goqu.L(`"cluster"::TEXT = 'local-cluster'`), goqu.L(`"data"->>'namespace'`).IsNull()).ToSQL()
+	if err != nil {
+		klog.Errorf("Error creating query [%s]. Error: [%+v]", query, err)
+		shared.csErr = err
+	}
+
+	rows, queryerr := cache.pool.Query(ctx, query)
+	if queryerr != nil {
+		klog.Errorf("Error resolving query [%s]. Error: [%+v]", query, queryerr.Error())
+		shared.csErr = queryerr
+		shared.csResources = []resource{}
+		return shared.csResources, shared.csErr
+	}
+
+	// var resource *resource
+	if rows != nil {
+		defer rows.Close()
+
+		for rows.Next() {
+			var kind string
+			var apigroup string
+			err := rows.Scan(&apigroup, &kind)
+			if err != nil {
+				klog.Errorf("Error %s retrieving rows for query:%s for apigroup %s and kind %s", err.Error(), query,
+					apigroup, kind)
+				shared.csErr = err
+				continue
```

Comment:
If we get error for one row, is this an error state ? - If we treat this as an error state we can break the loop - if we want to still use it , we have to remove the error right ?

Reply:
Okay since we reset the error in line 85 `err := rows.Scan(&apigroup, &kind)` we can set `shared.csErr = err` if err != nil  - in line 92.

Reply:
If we set the error - Do you still want to continue the loop OR ee can break ? 
Currently we can set an error , but we still populate the data from DB .

Reply:
the original idea was - if only one row has an error we don't want to exit the loop we still want to populate data and continue loop. Based on discussion, we can change the error to a log instead since we don't use that specific error for anything other than looking at the specific kind/apigroup that caused query to fail.